Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
+    convert :: PackageDB -> String
+    convert  GlobalPackageDB         = "-global-package-db"
+    convert  UserPackageDB           = "-user-package-db"
+    convert (SpecificPackageDB path) = "-package-db=" ++ path
 
-cabal_bench :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-cabal_bench config spec extraArgs = do
-    res <- cabal config spec [] ("bench" : extraArgs)
-    let r = recordRun res BenchSuccess nullResult
-    record spec r
-    return r
+------------------------------------------------------------------------
+-- * Running ghc-pkg
+
+ghcPkg :: String -> [String] -> TestM Result
+ghcPkg cmd args = do
+    db_path <- sharedDBPath
+    (config, test) <- ask
+    unless (testPackageDb test) $
+        error "Must initialize package database using withPackageDb"
+    let db_stack = packageDBStack config ++ [SpecificPackageDB db_path]
+        extraArgs = ghcPkgPackageDBParams db_stack
+    run Nothing (ghcPkgPath config) (cmd : extraArgs ++ args)
+
+ghcPkgPackageDBParams :: PackageDBStack -> [String]
+ghcPkgPackageDBParams dbs = concatMap convert dbs where
+    convert :: PackageDB -> [String]
+    -- TODO: this is a bit dodgy but there's no way good
+    -- way to handle this case in ghc-pkg
+    convert  GlobalPackageDB         = []
+    convert  UserPackageDB           = []
+    convert (SpecificPackageDB path) = ["--package-db=" ++ path]
 
-compileSetup :: SuiteConfig -> FilePath -> IO ()
-compileSetup config packageDir = do
-    r <- run (Just $ packageDir) (ghcPath config) []
-         [ "--make"
--- HPC causes trouble -- see #1012
---       , "-fhpc"
-         , "-package-conf " ++ (cabalDistPref config) </> "package.conf.inplace"
-         , "Setup.hs"
-         ]
-    requireSuccess r
+------------------------------------------------------------------------
+-- * Running other things
 
--- | Returns the command that was issued, the return code, and the output text.
-cabal :: SuiteConfig
-      -> PackageSpec
-      -> [(String, Maybe String)]  -- ^ environment variable overrides
-      -> [String]  -- ^ extra arguments
-      -> IO (String, ExitCode, String)
-cabal config spec envOverrides cabalArgs_ = do
-    let cabalArgs = case distPref spec of
-                       Nothing -> cabalArgs_
-                       Just dist -> ("--builddir=" ++ dist) : cabalArgs_
-    customSetup <- doesFileExist (directory spec </> "Setup.hs")
-    if customSetup
-        then do
-            compileSetup config (directory spec)
-            path <- canonicalizePath $ directory spec </> "Setup"
-            run (Just $ directory spec) path envOverrides cabalArgs
-        else do
-            -- Use shared Setup executable (only for Simple build types).
-            path <- canonicalizePath "Setup"
-            run (Just $ directory spec) path envOverrides cabalArgs
+-- | Running an executable that was produced by cabal.
+runExe :: String -> [String] -> TestM Result
+runExe exe_name args = do
```

Comment:
The naming convention seems to be a bit... inconsistent: runExe here, cabal_build above, etc. Perhaps it would be worth trying to bring it into line while you're rewriting everything anyway?